libpod/lock/shm/shm_lock.c

```diff
@@ -0,0 +1,383 @@
+// Returns a valid pointer on success or NULL on error.
+// If an error occurs, it will be written to the int pointed to by error_code.
+shm_struct_t *open_lock_shm(uint32_t num_locks, int *error_code) {
+  int shm_fd;
+  shm_struct_t *shm;
+  size_t shm_size;
+  uint32_t num_bitmaps;
+
+  if (error_code == NULL) {
+    return NULL;
+  }
+
+  // We need a nonzero number of locks
+  if (num_locks == 0) {
+    *error_code = EINVAL;
+    return NULL;
+  }
+
+  // Calculate the number of bitmaps required
+  if (num_locks % BITMAP_SIZE != 0) {
+    // Number of locks not a multiple of BITMAP_SIZE
+    *error_code = EINVAL;
+    return NULL;
+  }
+  num_bitmaps = num_locks / BITMAP_SIZE;
+
+  // Calculate size of the shm segment
+  shm_size = compute_shm_size(num_bitmaps);
+
+  shm_fd = shm_open(SHM_NAME, O_RDWR, 0600);
+  if (shm_fd < 0) {
+    return NULL;
+  }
+
+  // Map the shared memory in
+  shm = mmap(NULL, shm_size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd, 0);
+  if (shm == MAP_FAILED) {
+    *error_code = errno;
+  }
+
+  // Ignore errors, it's ok if we leak a single FD since this only runs once
+  close(shm_fd);
+
+  // Check if we successfully mmap'd
+  if (shm == MAP_FAILED) {
+    return NULL;
+  }
+
+  // Need to check the SHM to see if it's actually our locks
+  if (shm->magic != MAGIC) {
+    *error_code = errno;
+    goto CLEANUP;
+  }
+  if (shm->num_locks != num_locks) {
+    *error_code = errno;
+    goto CLEANUP;
+  }
+
+  return shm;
+
+ CLEANUP:
+  munmap(shm, shm_size);
+  return NULL;
+}
+
+// Close an open SHM lock struct, unmapping the backing memory.
+// The given shm_struct_t will be rendered unusable as a result.
+// On success, 0 is returned. On failure, negative ERRNO values are returned.
+int32_t close_lock_shm(shm_struct_t *shm) {
+  int ret_code;
+  size_t shm_size;
+
+  // We can't unmap null...
+  if (shm == NULL) {
+    return -1 * EINVAL;
+  }
+
+  shm_size = compute_shm_size(shm->num_bitmaps);
+
+  ret_code = munmap(shm, shm_size);
+
+  if (ret_code != 0) {
+    return -1 * errno;
+  }
+
+  return 0;
+}
+
+// Allocate the first available semaphore
+// Returns a positive integer guaranteed to be less than UINT32_MAX on success,
+// or negative errno values on failure
+// On sucess, the returned integer is the number of the semaphore allocated
+int64_t allocate_semaphore(shm_struct_t *shm) {
+  int ret_code, i;
+  bitmap_t test_map;
+  int64_t sem_number, num_within_bitmap;
+
+  if (shm == NULL) {
+    return -1 * EINVAL;
+  }
+
+  // Lock the semaphore controlling access to our shared memory
+  do {
+    ret_code = sem_wait(&(shm->segment_lock));
+  } while(ret_code == EINTR);
+  if (ret_code != 0) {
+    return -1 * errno;
+  }
+
+  // Loop through our bitmaps to search for one that is not full
+  for (i = 0; i < shm->num_bitmaps; i++) {
+    if (shm->locks[i].bitmap != 0xFFFFFFFF) {
+      test_map = 0x1;
+      num_within_bitmap = 0;
+      while (test_map != 0) {
```

Comment:
Absolutely non-blocking, this works fine:

I wouldn’t bother with two variables; just use `for (j = 0; j < BITMAP_SIZE; j++) … (1 << j)`.

- It is going to be more readable
- Compilers are pretty good at strength reduction, turning the `1 << index` into a variable shifted by 1 (and at judging whether it is an improvement at all)
- OTOH the loop termination by shift wraparound is a bit of a corner case that may not be understood.  A trivial for min/max/++ loop will definitely be understood.
- The loop in `deallocate_semaphore` to compute `test_map`, which exists pretty much only to mirror this one, could become a single operation as well.

(It’s pretty likely that gcc can recognize all of that, and both C implementations will end up generating the same code in the end, maybe even using specialized scan instructions. Given that we are writing almost everything in Golang, we are certainly not optimizing at this level, so I’d prefer simpler code.)